agent/grpc-internal/balancer/balancer.go

```diff
@@ -0,0 +1,184 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// NOTICE: This file is a derivative of grpc's grpc/pickfirst.go implementation.
+
+package balancer
+
+import (
+	"errors"
+	"fmt"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/grpclog"
+)
+
+// CustomBalancerName is the name of the modified pick_first balancer.
+const CustomBalancerName = "custom"
+
+var logger = grpclog.Component("balancer")
+
+func newCustomerBalancerBuilder() balancer.Builder {
+	return &customBalancerBuilder{}
+}
+
+type customBalancerBuilder struct{}
+
+func (*customBalancerBuilder) Build(cc balancer.ClientConn, opt balancer.BuildOptions) balancer.Balancer {
+	return &customBalancer{cc: cc}
+}
+
+func (*customBalancerBuilder) Name() string {
+	return CustomBalancerName
+}
+
+type customBalancer struct {
+	state   connectivity.State
+	cc      balancer.ClientConn
+	subConn balancer.SubConn
+}
+
+func (b *customBalancer) ResolverError(err error) {
+	if logger.V(2) {
+		logger.Infof("customBalancer: ResolverError called with error %v", err)
+	}
+	if b.subConn == nil {
+		b.state = connectivity.TransientFailure
+	}
+
+	if b.state != connectivity.TransientFailure {
+		// The picker will not change since the balancer does not currently
+		// report an error.
+		return
+	}
+	b.cc.UpdateState(balancer.State{
+		ConnectivityState: connectivity.TransientFailure,
+		Picker:            &picker{err: fmt.Errorf("name resolver error: %v", err)},
+	})
+}
+
+func (b *customBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+	if len(state.ResolverState.Addresses) == 0 {
+		// The resolver reported an empty address list. Treat it like an error by
+		// calling b.ResolverError.
+		b.ResolverError(errors.New("produced zero addresses"))
+		return balancer.ErrBadResolverState
+	}
+
+	// Unlike the pick_first balancer, always tear down existing subConn
+	if b.subConn != nil {
+		b.cc.RemoveSubConn(b.subConn)
+		b.subConn = nil
+	}
```

Comment:
This should be the only logical diff from https://github.com/grpc/grpc-go/blob/v1.49.0/pickfirst.go

Reply:
```diff
--- balancer.go	2022-12-06 18:41:15.000000000 -0500
+++ /Users/chriskim/go/pkg/mod/google.golang.org/grpc@v1.49.0/pickfirst.go	2022-10-27 12:43:02.000000000 -0400
@@ -16,9 +16,7 @@
  *
  */
 
-// NOTICE: This file is a derivative of grpc's grpc/pickfirst.go implementation.
-
-package balancer
+package grpc
 
 import (
 	"errors"
@@ -26,37 +24,34 @@
 
 	"google.golang.org/grpc/balancer"
 	"google.golang.org/grpc/connectivity"
-	"google.golang.org/grpc/grpclog"
 )
 
-// CustomBalancerName is the name of the modified pick_first balancer.
-const CustomBalancerName = "custom"
-
-var logger = grpclog.Component("balancer")
+// PickFirstBalancerName is the name of the pick_first balancer.
+const PickFirstBalancerName = "pick_first"
 
-func newCustomerBalancerBuilder() balancer.Builder {
-	return &customBalancerBuilder{}
+func newPickfirstBuilder() balancer.Builder {
+	return &pickfirstBuilder{}
 }
 
-type customBalancerBuilder struct{}
+type pickfirstBuilder struct{}
 
-func (*customBalancerBuilder) Build(cc balancer.ClientConn, opt balancer.BuildOptions) balancer.Balancer {
-	return &customBalancer{cc: cc}
+func (*pickfirstBuilder) Build(cc balancer.ClientConn, opt balancer.BuildOptions) balancer.Balancer {
+	return &pickfirstBalancer{cc: cc}
 }
 
-func (*customBalancerBuilder) Name() string {
-	return CustomBalancerName
+func (*pickfirstBuilder) Name() string {
+	return PickFirstBalancerName
 }
 
-type customBalancer struct {
+type pickfirstBalancer struct {
 	state   connectivity.State
 	cc      balancer.ClientConn
 	subConn balancer.SubConn
 }
 
-func (b *customBalancer) ResolverError(err error) {
+func (b *pickfirstBalancer) ResolverError(err error) {
 	if logger.V(2) {
-		logger.Infof("customBalancer: ResolverError called with error %v", err)
+		logger.Infof("pickfirstBalancer: ResolverError called with error %v", err)
 	}
 	if b.subConn == nil {
 		b.state = connectivity.TransientFailure
@@ -73,24 +68,29 @@
 	})
 }
 
-func (b *customBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+func (b *pickfirstBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
 	if len(state.ResolverState.Addresses) == 0 {
 		// The resolver reported an empty address list. Treat it like an error by
 		// calling b.ResolverError.
+		if b.subConn != nil {
+			// Remove the old subConn. All addresses were removed, so it is no longer
+			// valid.
+			b.cc.RemoveSubConn(b.subConn)
+			b.subConn = nil
+		}
 		b.ResolverError(errors.New("produced zero addresses"))
 		return balancer.ErrBadResolverState
 	}
 
-	// Unlike the pick_first balancer, always tear down existing subConn
 	if b.subConn != nil {
-		b.cc.RemoveSubConn(b.subConn)
-		b.subConn = nil
+		b.cc.UpdateAddresses(b.subConn, state.ResolverState.Addresses)
+		return nil
 	}
 
 	subConn, err := b.cc.NewSubConn(state.ResolverState.Addresses, balancer.NewSubConnOptions{})
 	if err != nil {
 		if logger.V(2) {
-			logger.Errorf("customBalancer: failed to NewSubConn: %v", err)
+			logger.Errorf("pickfirstBalancer: failed to NewSubConn: %v", err)
 		}
 		b.state = connectivity.TransientFailure
 		b.cc.UpdateState(balancer.State{
@@ -109,13 +109,13 @@
 	return nil
 }
 
-func (b *customBalancer) UpdateSubConnState(subConn balancer.SubConn, state balancer.SubConnState) {
+func (b *pickfirstBalancer) UpdateSubConnState(subConn balancer.SubConn, state balancer.SubConnState) {
 	if logger.V(2) {
-		logger.Infof("customBalancer: UpdateSubConnState: %p, %v", subConn, state)
+		logger.Infof("pickfirstBalancer: UpdateSubConnState: %p, %v", subConn, state)
 	}
 	if b.subConn != subConn {
 		if logger.V(2) {
-			logger.Infof("customBalancer: ignored state change because subConn is not recognized")
+			logger.Infof("pickfirstBalancer: ignored state change because subConn is not recognized")
 		}
 		return
 	}
@@ -149,10 +149,10 @@
 	}
 }
 
-func (b *customBalancer) Close() {
+func (b *pickfirstBalancer) Close() {
 }
 
-func (b *customBalancer) ExitIdle() {
+func (b *pickfirstBalancer) ExitIdle() {
 	if b.subConn != nil && b.state == connectivity.Idle {
 		b.subConn.Connect()
 	}
@@ -179,5 +179,5 @@
 }
 
 func init() {
-	balancer.Register(newCustomerBalancerBuilder())
+	balancer.Register(newPickfirstBuilder())
 }
```

Reply:
```diff
--- /Users/chriskim/go/pkg/mod/google.golang.org/grpc@v1.49.0/pickfirst.go	2022-10-27 12:43:02.000000000 -0400
+++ balancer.go	2022-12-06 18:41:15.000000000 -0500
@@ -16,7 +16,9 @@
  *
  */
 
-package grpc
+// NOTICE: This file is a derivative of grpc's grpc/pickfirst.go implementation.
+
+package balancer
 
 import (
 	"errors"
@@ -24,34 +26,37 @@
 
 	"google.golang.org/grpc/balancer"
 	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/grpclog"
 )
 
-// PickFirstBalancerName is the name of the pick_first balancer.
-const PickFirstBalancerName = "pick_first"
+// CustomBalancerName is the name of the modified pick_first balancer.
+const CustomBalancerName = "custom"
+
+var logger = grpclog.Component("balancer")
 
-func newPickfirstBuilder() balancer.Builder {
-	return &pickfirstBuilder{}
+func newCustomerBalancerBuilder() balancer.Builder {
+	return &customBalancerBuilder{}
 }
 
-type pickfirstBuilder struct{}
+type customBalancerBuilder struct{}
 
-func (*pickfirstBuilder) Build(cc balancer.ClientConn, opt balancer.BuildOptions) balancer.Balancer {
-	return &pickfirstBalancer{cc: cc}
+func (*customBalancerBuilder) Build(cc balancer.ClientConn, opt balancer.BuildOptions) balancer.Balancer {
+	return &customBalancer{cc: cc}
 }
 
-func (*pickfirstBuilder) Name() string {
-	return PickFirstBalancerName
+func (*customBalancerBuilder) Name() string {
+	return CustomBalancerName
 }
 
-type pickfirstBalancer struct {
+type customBalancer struct {
 	state   connectivity.State
 	cc      balancer.ClientConn
 	subConn balancer.SubConn
 }
 
-func (b *pickfirstBalancer) ResolverError(err error) {
+func (b *customBalancer) ResolverError(err error) {
 	if logger.V(2) {
-		logger.Infof("pickfirstBalancer: ResolverError called with error %v", err)
+		logger.Infof("customBalancer: ResolverError called with error %v", err)
 	}
 	if b.subConn == nil {
 		b.state = connectivity.TransientFailure
@@ -68,29 +73,24 @@
 	})
 }
 
-func (b *pickfirstBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+func (b *customBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
 	if len(state.ResolverState.Addresses) == 0 {
 		// The resolver reported an empty address list. Treat it like an error by
 		// calling b.ResolverError.
-		if b.subConn != nil {
-			// Remove the old subConn. All addresses were removed, so it is no longer
-			// valid.
-			b.cc.RemoveSubConn(b.subConn)
-			b.subConn = nil
-		}
 		b.ResolverError(errors.New("produced zero addresses"))
 		return balancer.ErrBadResolverState
 	}
 
+	// Unlike the pick_first balancer, always tear down existing subConn
 	if b.subConn != nil {
-		b.cc.UpdateAddresses(b.subConn, state.ResolverState.Addresses)
-		return nil
+		b.cc.RemoveSubConn(b.subConn)
+		b.subConn = nil
 	}
 
 	subConn, err := b.cc.NewSubConn(state.ResolverState.Addresses, balancer.NewSubConnOptions{})
 	if err != nil {
 		if logger.V(2) {
-			logger.Errorf("pickfirstBalancer: failed to NewSubConn: %v", err)
+			logger.Errorf("customBalancer: failed to NewSubConn: %v", err)
 		}
 		b.state = connectivity.TransientFailure
 		b.cc.UpdateState(balancer.State{
@@ -109,13 +109,13 @@
 	return nil
 }
 
-func (b *pickfirstBalancer) UpdateSubConnState(subConn balancer.SubConn, state balancer.SubConnState) {
+func (b *customBalancer) UpdateSubConnState(subConn balancer.SubConn, state balancer.SubConnState) {
 	if logger.V(2) {
-		logger.Infof("pickfirstBalancer: UpdateSubConnState: %p, %v", subConn, state)
+		logger.Infof("customBalancer: UpdateSubConnState: %p, %v", subConn, state)
 	}
 	if b.subConn != subConn {
 		if logger.V(2) {
-			logger.Infof("pickfirstBalancer: ignored state change because subConn is not recognized")
+			logger.Infof("customBalancer: ignored state change because subConn is not recognized")
 		}
 		return
 	}
@@ -149,10 +149,10 @@
 	}
 }
 
-func (b *pickfirstBalancer) Close() {
+func (b *customBalancer) Close() {
 }
 
-func (b *pickfirstBalancer) ExitIdle() {
+func (b *customBalancer) ExitIdle() {
 	if b.subConn != nil && b.state == connectivity.Idle {
 		b.subConn.Connect()
 	}
@@ -179,5 +179,5 @@
 }
 
 func init() {
-	balancer.Register(newPickfirstBuilder())
+	balancer.Register(newCustomerBalancerBuilder())
 }
```

Reply:
Just for the sake of asking the question: this is _slightly_ different to the current behaviour in that in the case of three servers, there is currently about a 33% chance that we don't have to tear down the conn and so can keep streams alive for another rebalance interval. This change makes sure that even if we randomly picked the same server as before, we still close all the streams and re-open them.

Now it's worst case only 33% more likely to interrupt a stream and with more servers that probability diminishes so I could easily be persuaded that the cost is low enough that most people won't notice and the simpler code is worth it... _but_ just for the sake of having the discussion, did you consider making it behave the same so that it checks if the first addr is different and if not keeps it? I think if that is easy (i.e. just an extra comparison in this if clause) then it might be worth it mostly for the sake of keeping externally observable behaviour identical more than the small-optimization.

Reply:
Ah I see why it's not super simple: the [`balancer.SubConn` interface](https://pkg.go.dev/google.golang.org/grpc/balancer#SubConn) doesn't let us easily figure out the actual address `b.subConn` is currently connected to.

We could store our own copy of the list of addrs and just assume it was successful in connecting to the first one (which is probably true 99.99% of the time) but it might actually have failed and so connected to another which means we might make a mistake and not close it even though it's currently connected to a different server than the new first one.

This is probably complex enough to be in the "let's leave it simpler and accept the slight change in behaviour" camp but I'd love to hear your thoughts! I don't think there is a way from either the `SubConn` or `ClientConn` interfaces to figure out which address is actually connected to sadly.

```go
	// Unlike the pick_first balancer, always tear down existing subConn
	if b.subConn != nil && b.subConn. {
		b.cc.RemoveSubConn(b.subConn)
		b.subConn = nil
	}
```

Reply:
Great points! Yes it was due to not having reference of the currently active connection.
I've pushed a refactor that gives us a reliable way of storing the active address so we can preserve the 1/N chance of keeping the connection alive. Check it out and let me know if it looks okay.